proknow-sdk/Upload/Uploads.cs

```diff
@@ -85,83 +86,95 @@ public async Task<IList<UploadResult>> UploadAsync(string workspace, string path
+                    // Remove any uploads that reached terminal status from the collection of unresolved uploads
+                    foreach (var resolvedUploadId in resolvedUploadIds)
+                    {
+                        var index = unresolvedUploadResults.FindIndex(t => t.Id == resolvedUploadId);
+                        unresolvedUploadResults.RemoveAt(index);
+                    }
 
-                    // If one was found
-                    if (lastTerminalUploadProcessingResult != null)
+                    // If there are still unresolved uploads and possibly another page of results
+                    if (unresolvedUploadResults.Count > 0 && thisPageUploadProcessingResults.Count > 0)
                     {
-                        // Update the query parameters to search for processed uploads updated after this one
-                        if (queryParameters == null)
+                        // Get the last upload on the current page whose processing has reached a terminal status
+                        var lastTerminalUploadProcessingResult = thisPageUploadProcessingResults.LastOrDefault(t => _terminalStatuses.Contains(t.Status));
+
+                        // If one was found
+                        if (lastTerminalUploadProcessingResult != null)
                         {
-                            queryParameters = new Dictionary<string, object>();
+                            // Update the query parameters to search for processed uploads updated after this one
+                            if (queryParameters == null)
+                            {
+                                queryParameters = new Dictionary<string, object>();
+                            }
+                            queryParameters["updated"] = lastTerminalUploadProcessingResult.UpdatedAt;
+                            queryParameters["after"] = lastTerminalUploadProcessingResult.Id;
                         }
-                        queryParameters["updated"] = lastTerminalUploadProcessingResult.UpdatedAt;
-                        queryParameters["after"] = lastTerminalUploadProcessingResult.Id;
                     }
+                    else
+                    {
+                        break;
+                    }
+                } // End loop through pages of processing results
 
-                    // Give the uploads some time to process with each retry waiting twice as long as the previous retry
-                    await Task.Delay(nextRetryDelay);
-                    totalRetryDelay += nextRetryDelay;
-                    nextRetryDelay *= 2;
+                // If there are still unresolved uploads and retries have not been exhausted
+                if (unresolvedUploadResults.Count > 0 && retryDelayIndex < RetryDelays.Count)
+                {
+                    // Give the uploads some time to process
+                    await Task.Delay(RetryDelays[retryDelayIndex++]);

```

Comment:
Might it be a good idea to reset the retryDelayIndex back to 0 if you end up finding a new upload update on a subsequent iteration?

Reply:
I'll fix this.